File: raspi_tank/src/motor.erl
-module(motor).
-author('josh@isotope11.com').

-export([init/1, set_speed/2, pi_blast/1]).

-record(pin, {pin_number, value = 0}).

init(Motor) ->
  #pin{pin_number = Motor}.

%% set_speed takes a number between -1 and 1 and will map it to
%% the appropriate values for a Sabretooth RC.

set_speed(Value, Pin) ->
  NewValue = mapped_value_for(Value),
  Pin#pin{value=NewValue}.

pi_blast(#pin{pin_number=PinNumber, value=Value}) ->
  pi_blaster:set_pin(PinNumber, Value).

% Maps a value from -1 to 1 to the appropriate value to drive a
% Sabretooth RC, assuming 1 is all the way forward and -1 is all
% the way backwards.
mapped_value_for(Value) ->
  idle() + (Value * 500)/max_pulse_width().

idle() ->
  1500 / max_pulse_width().

max_pulse_width() ->
  10000.

File: elixir/raspi_tank/lib/raspi_tank/motor.ex
defmodule RaspiTank.Motor do
  defrecord Pin, number: nil, value: nil

  def init(pin), do: Pin[number: pin]

  def set_speed(value, pin) do
    mapped_value_for(value) |> pin.value
  end

  def pi_blast(Pin[number: number, value: value]) do
    PiBlaster.set_pin(number, value)
  end

  defp mapped_value_for(value) do
    idle + (value * 500)/max_pulse_width
  end

  defp idle, do: 1_500 / max_pulse_width
  defp max_pulse_width, do: 10_000
end
